webapp/src/test/java/com/box/l10n/mojito/smartling/SmartlingOAuth2TokenServiceTest.java

```diff
@@ -0,0 +1,134 @@
+    MockitoAnnotations.initMocks(this);
+    smartlingOAuth2TokenService =
+        new SmartlingOAuth2TokenService(
+            "clientId",
+            "clientSecret",
+            "https://test.local/auth",
+            "https://test.local/auth/refresh");
+    smartlingOAuth2TokenService.client = httpClient;
+  }
+
+  @Test
+  public void getAccessTokenRequestsNewTokenWhenRefreshTokenExpired() throws Exception {
+    HttpResponse<String> httpResponse =
+        createMockResponse(
+            "{\"response\": {\"data\": {\"accessToken\": \"newToken\", \"expiresIn\": 3600, \"refreshToken\": \"refreshToken\", \"refreshExpiresIn\": 1}}}");
+    when(httpClient.send(any(HttpRequest.class), any(HttpResponse.BodyHandler.class)))
+        .thenReturn(httpResponse);
+
+    smartlingOAuth2TokenService.getAccessToken(); // First call to get the token
+    String accessToken =
+        smartlingOAuth2TokenService.getAccessToken(); // Second call should request a new token
+
+    assertEquals("newToken", accessToken);
+    verify(httpClient, times(2)).send(any(), any());
+  }
+
+  @Test
+  public void getAccessTokenRequestsNewTokenWhenNoTokenExists() throws Exception {
+    HttpResponse<String> httpResponse =
+        createMockResponse(
+            "{\"response\": {\"data\": {\"accessToken\": \"newToken\", \"expiresIn\": 3600, \"refreshToken\": \"refreshToken\", \"refreshExpiresIn\": 7200}}}");
+    when(httpClient.send(any(HttpRequest.class), any(HttpResponse.BodyHandler.class)))
+        .thenReturn(httpResponse);
+
+    String accessToken = smartlingOAuth2TokenService.getAccessToken();
+
+    assertEquals("newToken", accessToken);
+    verify(httpClient, times(1)).send(any(HttpRequest.class), any(HttpResponse.BodyHandler.class));
+  }
+
+  @Test
+  public void getRefreshedAccessTokenTokenWhenAccessTokenExpired() throws Exception {
+    HttpResponse<String> httpResponse =
+        createMockResponse(
+            "{\"response\": {\"data\": {\"accessToken\": \"newToken\", \"expiresIn\": 1, \"refreshToken\": \"refreshToken\", \"refreshExpiresIn\": 7200}}}");
+    when(httpClient.send(any(HttpRequest.class), any(HttpResponse.BodyHandler.class)))
+        .thenReturn(httpResponse);
+
+    String accessToken = smartlingOAuth2TokenService.getAccessToken();
+    assertEquals("newToken", accessToken);
+    httpResponse =
+        createMockResponse(
+            "{\"response\": {\"data\": {\"accessToken\": \"refreshedToken\", \"expiresIn\": 1, \"refreshToken\": \"refreshToken\", \"refreshExpiresIn\": 7200}}}");
+    when(httpClient.send(any(HttpRequest.class), any(HttpResponse.BodyHandler.class)))
+        .thenReturn(httpResponse);
+    Thread.sleep(2000);
+    accessToken = smartlingOAuth2TokenService.getAccessToken();
+
+    assertEquals("refreshedToken", accessToken);
+    verify(httpClient, times(2)).send(any(HttpRequest.class), any(HttpResponse.BodyHandler.class));
+  }
+
+  private HttpResponse<String> createMockResponse(String body) {
```

Comment:
can't do that with mockito?

Reply:
I was having issues with Mockito on Friday, appears the problem was I wasn't declaring the generic type of the mock e.g. `@Mock HttpResponse` rather than `@Mock HttpResponse<String>` 🤦